app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt

```diff
@@ -51,10 +55,32 @@ class ThreadViewModel : ViewModel() {
     }
 
     fun loadMessages(threadUid: String) {
-        val thread = MailboxContentController.getThread(threadUid) ?: return
-        MailData.selectThread(thread)
-        thread.markAsSeen()
-        MailData.loadMessages(thread)
+        MailboxContentController.getThread(threadUid)?.let { thread ->
+            MailData.selectThread(thread)
+            markAsSeen(thread)
+            MailData.loadMessages(thread)
+        }
+    }
+
+    private fun markAsSeen(thread: Thread) {
+        if (thread.unseenMessagesCount != 0) {
+
+            val mailboxUuid = MailData.currentMailboxFlow.value?.uuid ?: return
+
+            MailRealm.mailboxContent.writeBlocking {
+                getLatestThread(thread.uid)?.let { latestThread ->
+
+                    val apiResponse = ApiRepository.markMessagesAsSeen(mailboxUuid, latestThread.messages.map { it.uid })
+
+                    if (apiResponse.isSuccess()) {
+                        latestThread.apply {
+                            messages.forEach { it.seen = true }
+                            unseenMessagesCount = 0
+                        }
+                    }
+                }
+            }
```

Comment:
Move realm operation in its controller.
In addition we can make the call api well before because here we are sure that the thread exists in realm